mathics/builtin/manipulate.py

```diff
@@ -0,0 +1,314 @@
+
+try:
+    from ipywidgets import (IntSlider, FloatSlider, ToggleButtons, Box, DOMWidget)
+    _ipywidgets = True
+except ImportError:
+    # fallback to non-Manipulate-enabled build if we don't have ipywidgets installed.
+    _ipywidgets = False
+
+
+"""
+A basic implementation of Manipulate[]. There is currently no support for Dynamic[] elements.
+This implementation is basically a port from ipywidget.widgets.interaction for Mathics.
+"""
+
+def _interactive(interact_f, kwargs_widgets):
+    # this is a modified version of interactive() in ipywidget.widgets.interaction
+
+    container = Box(_dom_classes=['widget-interact'])
+    container.children = [w for w in kwargs_widgets if isinstance(w, DOMWidget)]
+
+    def call_f(name=None, old=None, new=None):
+        kwargs = dict((widget._kwarg, widget.value) for widget in kwargs_widgets)
+        try:
+            interact_f(**kwargs)
+        except Exception as e:
+            container.log.warn("Exception in interact callback: %s", e, exc_info=True)
+
+    for widget in kwargs_widgets:
+        widget.on_trait_change(call_f, 'value')
+
+    container.on_displayed(lambda _: call_f(None, None, None))
+
+    return container
+
+
+def _strip_namespace(name):
+    return name.split('`')[-1]  # e.g. "Global`x" will turn into "x"
+
+
+def _manipulate_label(x):
+    if isinstance(x, String):
+        return x.get_string_value()
+    elif isinstance(x, Symbol):
+        return _strip_namespace(x.get_name())
+    else:
+        return str(x)
+
+
+class PatternDispatcher:
+    '''A pattern matcher similar to the one for BuiltIns, but usable in any context (e.g. locally)'''
+
+    def __init__(self):
+        from mathics.core.parser import parse_builtin_rule
+        from mathics.builtin.patterns import Pattern
+
+        self._patterns = []
+        for name in dir(self):
+            if name.startswith('apply_'):
+                function = getattr(self, name)
+                self._patterns.append((Pattern.create(
+                    parse_builtin_rule(function.__doc__)), function))
+
+    def dispatch(self, expr, evaluation):
+        from mathics.builtin.patterns import StopGenerator
+
+        class StopGenerator_MatchQ(StopGenerator):
+            pass
+
+        try:
+            for pattern, function in self._patterns:
+                def yield_func(vars, rest):
+                    function(**dict((_strip_namespace(key), value) for (key, value) in vars.items()))
+                    raise StopGenerator_MatchQ(Symbol("True"))
+
+                pattern.match(yield_func, expr, {}, evaluation)
+        except StopGenerator_MatchQ:
+            return True
+
+        return False
+
+
+class IllegalWidgetArguments(Exception):
+    def __init__(self, var):
+        super(IllegalWidgetArguments, self).__init__()
+        self.var = var
+
+
+class Manipulations(PatternDispatcher):
+    def __init__(self, evaluation):
+        if six.PY2:
+            PatternDispatcher.__init__(self)
```

Comment:
I don't think we need this (even under Python 2) if `PatternDispatcher` inherits from `object`